flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleanerTest.java

```diff
@@ -181,33 +212,97 @@ public void testMediumPriorityCleanupBlocksAllLowerPrioritizedCleanups() {
+        assertThat(oneRunCleanup.isDone()).isTrue();
+        assertThat(actualJobIds).containsExactly(JOB_ID, JOB_ID);
+    }
+
     private static DefaultResourceCleaner.Builder<CleanupCallback> createTestInstanceBuilder() {
+        return createTestInstanceBuilder(new FixedRetryStrategy(0, Duration.ZERO));
+    }
+
+    private static DefaultResourceCleaner.Builder<CleanupCallback> createTestInstanceBuilder(
+            RetryStrategy retryStrategy) {
         return DefaultResourceCleaner.forCleanableResources(
                 ComponentMainThreadExecutorServiceAdapter.forMainThread(),
                 EXECUTOR,
-                CleanupCallback::cleanup);
+                CleanupCallback::apply,
+                retryStrategy);
+    }
+
+    private static CleanupCallback cleanupWithRetry(
+            Collection<JobID> actualJobIds, int numberOfFailureRuns) {
+        final AtomicInteger failureRunCount = new AtomicInteger(numberOfFailureRuns);
+        return (actualJobId, executor) -> {
+            actualJobIds.add(actualJobId);
+            if (failureRunCount.getAndDecrement() > 0) {
+                return FutureUtils.completedExceptionally(
+                        new RuntimeException("Expected RuntimeException"));
+            }
+
+            return FutureUtils.completedVoidFuture();
+        };
+    }
+
+    private interface CleanupCallback extends BiFunction<JobID, Executor, CompletableFuture<Void>> {
+        // empty interface to remove necessity use generics all the time
     }
 
-    private static class CleanupCallback {
+    private static class SingleCallCleanup implements CleanupCallback {
 
         private final CompletableFuture<Void> resultFuture = new CompletableFuture<>();
         private JobID jobId;
 
         private final Consumer<CompletableFuture<Void>> internalFunction;
 
-        public static CleanupCallback withCompletionOnCleanup() {
-            return new CleanupCallback(resultFuture -> resultFuture.complete(null));
+        public static SingleCallCleanup withCompletionOnCleanup() {
+            return new SingleCallCleanup(resultFuture -> resultFuture.complete(null));
         }
 
-        public static CleanupCallback withoutCompletionOnCleanup() {
-            return new CleanupCallback(ignoredResultFuture -> {});
+        public static SingleCallCleanup withoutCompletionOnCleanup() {
+            return new SingleCallCleanup(ignoredResultFuture -> {});
         }
 
-        private CleanupCallback(Consumer<CompletableFuture<Void>> internalFunction) {
+        private SingleCallCleanup(Consumer<CompletableFuture<Void>> internalFunction) {
             this.internalFunction = internalFunction;
         }
 
-        public CompletableFuture<Void> cleanup(JobID jobId, Executor executor) {
+        public CompletableFuture<Void> apply(JobID jobId, Executor executor) {
             Preconditions.checkState(this.jobId == null);
             this.jobId = jobId;
```

Comment:
Would it make sense to wrap this in a synchronized, to avoid a race condition?

Reply:
Yes, the contract is that the method's body is run from within the main thread (see [DefaultResourceCleaner:127](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleaner.java#L127). Hence, all the synchronization happens through the main thread task queue. The test implementation does not add any concurrency.